composer/trainer/trainer.py

```diff
@@ -680,25 +685,44 @@ def _train_loop(self) -> None:
                         "trainer/global_step": self.state.step,
                         "trainer/batch_idx": self.state.timer.batch_in_epoch.value,
                     })
+                    # Rerun train_batch with smaller mini-batches if we hit CUDA out of memory
+                    rerun_train_batch = True
                     total_loss = None
-                    microbatches = self._train_data_spec.split_batch(state.batch, state.grad_accum)
-                    if self.deepspeed_enabled:
-                        total_loss = self._train_batch(microbatches)
-                    elif self._use_closures():
-                        for optimizer in state.optimizers:
-                            if use_grad_scaling:
-                                total_loss = state.scaler.step(
-                                    optimizer, closure=lambda **kwargs: self._train_batch(microbatches, **kwargs))
+                    while rerun_train_batch:
```

Comment:
Code style-wise, this is a lot of indentation now. Can you refactor some of the inner code out into a new function?

Reply:
Yep, will move the whole while loop into a new fn